src/main/java/org/nixos/idea/settings/NixLangSettingsConfigurable.java

```diff
@@ -0,0 +1,107 @@
+package org.nixos.idea.settings;
+
+import com.intellij.openapi.options.Configurable;
+import com.intellij.openapi.options.ConfigurationException;
+import com.intellij.openapi.options.SearchableConfigurable;
+import com.intellij.openapi.util.NlsContexts;
+import com.intellij.ui.RawCommandLineEditor;
+import com.intellij.ui.TitledSeparator;
+import com.intellij.ui.components.JBCheckBox;
+import com.intellij.ui.components.JBLabel;
+import com.intellij.util.ui.FormBuilder;
+import org.jetbrains.annotations.NonNls;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.nixos.idea.lsp.ui.CommandSuggestionsPopup;
+
+import javax.swing.*;
+import java.util.List;
+
+public class NixLangSettingsConfigurable implements SearchableConfigurable, Configurable.Beta {
```

Comment:
note: You could also use Kotlin for this class if you like. It may help because it allows you to use [Kotlin UI DSL 2](https://plugins.jetbrains.com/docs/intellij/kotlin-ui-dsl-version-2.html). I started to use it at https://github.com/NixOS/nix-idea/pull/79/files#diff-6e23a87822c87dd6d0da6e416237368c2f7f1f207110691e073fbf18d502be98 (not yet merged), just in case you want to take a lock. However, I have not yet looked into how to combine that with the `CommandSuggestionsPopup`.

Reply:
On a related note, if you are hoping to migrate the plugin to Kotlin entirely, I am happy to help in separate PRs!

Reply:
Not really, At least not yet. I think the language has its pros and cons. While I will not reject any new feature implemented in Kotlin, I am still a bit on defense about the language. While it has many nice features, I also feel it tends to encourage some bad patterns. At least what I consider bad patterns. Examples would be excessive use of extension functions and top-level methods. So, I think I will probably experiment with Kotlin a bit at selected places, before I would replace/rewrite existing logic.

Reply:
fair enough! I am partial to Kotlin, especially for more concurrency-heavy code, but it does take  discipline. Having worked with both though, my experience is it can be error-prone to mix with Java.

They have very different code structure philosphies, and the type system can trip you up nullability and concurrency wise, so beware with keeping both around. It might be more sensible to stick to Java until you feel comfortable making the jump (if at all).

Reply:
on that spirit, I prefer keeping this code in Java for consistency. I also feel like JB's Kolin UI DSL is not mature enough and requires using Java classes anyway (for the Command box for example)

Reply:
> my experience is it can be error-prone to mix with Java.

I am only aware of potential issues when nullability when you call Java-code from Kotlin. Especially when using Generics, as you can see at gradle/gradle#12388. I haven't noticed any other problems yet with my rather limited experience. At least I didn't have any problems so far when calling Kotlin code of IDEA from Java. Of course, the coroutines and some other constructs don't exist in Java and cannot be used when Java is involved. Do you have any specific hints about potential problems?

> you will find yourself writing more small data classes and top-level functions - instead of static members and abstract classes

I think I use a lot of small data classes in Java already and try to avoid abstract classes. Regarding top-level functions: I think it is really nice in Java that you have a namespace matching the file name attached to all methods and other constructs. It allows you to identify the file where the construct is defined. (At least as long as you are following the practice of naming the file according to the class, which is very common practice and partially enforced by the compiler.) I regularly struggle to find the correct file when looking at Kotlin code at GitHub. I don't think that I will ever be converted to top-level functions, assuming I haven't missed some obvious solution to look up the file which defines a method without searching through other files. 😄 Before I worked with Java, I used C++, which has the same problem that you cannot identify the source of a symbol without tooling, which I also always find quite annoying.

Reply:
> Do you have any specific hints about potential problems?

the painful bits imo come from using non-Kotlin frameworks
- serialisation libs don't know about Kotlin nullability and sealed classes
- Java collections are seen as 'read-only' from Kotlin (but Java libs can still mutate them)
- like you said, anything backend-related will not let you use coroutines. Chances are you will start putting Java-blocking operations into your coroutine dispatchers and unepxectedly blocking all threads in a pool
- when working with legacy, a lot of Java code is not annotated, making nullability a guessing game
- for this plugin specifically, expect IDEA APIs (which are originally Java) to not be complete in pure Kotlin. You then have to go investigating how to revert back to using old Java classes (usually imperative) within modern Kotlin API (usually declarative).

Ultimately the issues you bump into will probably depend on the stack you work on! I remember suffering with JOOQ too, before they had a Kotlin API.

> I regularly struggle to find the correct file when looking at Kotlin code 

That's fair. I have gotten used to using Ctrl+F and `/`-shortuct in github to find symbols. I now work in Go which also encourages top-level functions a lot, and Nix also has this tendency.

If you are curious, I encourage you to try out a pure-Kotlin project to see how it goes :)